### PR TITLE
Relax low resource utlization upper limit for resource distribution goals

### LINKED ISSUE #1416: Improvement on the low utilization threshold
Currently, for resource distribution goals, if average resource utilization is under a user-specified `lowUtilizationThreshold`, the resource balance lower and upper limits are calculated basing on the `lowUtilizationThreshold` (see [this PR](https://github.com/linkedin/cruise-control/pull/1391) for implementation details).

However, in a scenario that the `averageResourceUtlization` is a tiny bit smaller than the `lowUtilizationThreshold` (e.g. `9.999%` < `10%`), the resource distribution goal to be typically violated since many brokers will have resource utilization greater than the `lowUtilizationThreshold` given the fact the average is already so close the upper limit.

In order to mitigate the issue described above (when the `averageResourceUtlization` is a tiny bit smaller than the `lowUtilizationThreshold`), one approach is to relax the `lowUtilizationThreshold` by setting it to the max of the below two values:

- user-specified `lowUtilizationThreshold`
- the earlier percentage-based upper limit

### PR DESCRIPTION
**Pull request by Lincong**
This PR resolves #1416.

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtils.java [hunk 1]
```diff
@@ -530,7 +530,10 @@ public static double computeResourceUtilizationBalanceThreshold(double avgUtiliz
 
     } else {
       if (isLowUtilization) {
-        return balancingConstraint.lowUtilizationThreshold(resource) * balanceMargin;
+        double resourceBalanceUtilizationUpperLimit =
+            Math.max(balancingConstraint.resourceBalancePercentage(resource) * avgUtilizationPercentage,
+                     balancingConstraint.lowUtilizationThreshold(resource));
+        return resourceBalanceUtilizationUpperLimit * balanceMargin;
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtils.java:536]: This looks incorrect to me.
Shouldn't we pick the maximum between the following:
1. `balancingConstraint.lowUtilizationThreshold(resource) * balanceMargin` and 
2. `avgUtilizationPercentage * (1 + balancePercentageWithMargin)`

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtilsTest.java [hunk 2]
```diff
@@ -13,34 +13,51 @@
...
 
-    // Verify 4: Not low utilization and compute balance threshold upper bound
+    // Verify case 5: Not low utilization and compute balance threshold upper bound
     verifyComputingResourceUtilizationBalanceThreshold(resource, false, 0.2, 0.3 * (1 + ((1.3 * 1.2) - 1) * ResourceDistributionGoal.BALANCE_MARGIN));
   }
 
+  private void verifyComputingResourceUtilizationBalanceThreshold(Resource resource,
+      boolean isLowerThreshold,
+      double lowUtilizationThreshold,
+      double expectedComputedBalanceThreshold) {
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtilsTest.java:44]: Nit: Bad indentation

**Review by efeg — CHANGES_REQUESTED** [review]
Thanks for the patch -- left some comments.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtils.java (hunk 1 above)
Comment (Lincong) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtils.java:536]: Right. This way does make more sense

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtilsTest.java (hunk 2 above)
Comment (Lincong) [reply cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtilsTest.java:44]: Fixex

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtilsTest.java [hunk 3]
```diff
@@ -40,7 +46,7 @@ private void verifyComputingResourceUtilizationBalanceThreshold(Resource resourc
 
     BalancingConstraint mockBalanceConstraint = EasyMock.mock(BalancingConstraint.class);
     EasyMock.expect(mockBalanceConstraint.lowUtilizationThreshold(resource)).andReturn(lowUtilizationThreshold).anyTimes();
-    EasyMock.expect(mockBalanceConstraint.resourceBalancePercentage(resource)).andReturn(RESOURCE_BALANCE_THRESHOLD).anyTimes();
+    EasyMock.expect(mockBalanceConstraint.resourceBalancePercentage(resource)).andReturn(DEFAULT_RESOURCE_BALANCE_THRESHOLD).anyTimes();
     EasyMock.expect(mockBalanceConstraint.goalViolationDistributionThresholdMultiplier()).
         andReturn(GOAL_VIOLATION_DISTRIBUTION_THRESHOLD_MULTIPLIER).anyTimes();
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtilsTest.java:52]: Nit: (not introduced in this patch): Bad indentation
```
    double computedBalanceThreshold = GoalUtils.computeResourceUtilizationBalanceThreshold(AVG_UTILIZATION_PERCENTAGE,
                                                                                           resource,
                                                                                           mockBalanceConstraint, 
                                                                                           true,
                                                                                           ResourceDistributionGoal.BALANCE_MARGIN,
                                                                                           isLowerThreshold);
```

**Review by efeg — APPROVED** [review]
Thanks for the patch!
Left a nit comment, but feel free to ignore it. LGTM.

**Pull request merged**